### PR TITLE
[Executor] Support applying default value and ensuring type for flex flow

### PR DESCRIPTION
**Pull request by lumoslnt**
# Description

In this PR, we support applying default value and ensuring type provided for flex flow. If the input is not provided then we will apply the default value defined in the yaml, also we will check whether the input type is consistent with the type defined in yaml.  An example yaml file:
```
entry: my_flow:MyClass
init:
  input_init:
    type: string
    default: input_init
inputs:
  input_1:
    type: string
  input_2:
    type: string
    default: input_2
```

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/executor/_script_executor.py [hunk 1]
```diff
@@ -445,3 +452,15 @@ def _parse_flow_file(self):
...
+    def _init_input_sign(self):
+        if not self.is_function_entry:
+            with open(self._working_dir / self._flow_file, "r", encoding="utf-8") as fin:
+                flow_dag = load_yaml(fin)
+            flow = FlexFlow.deserialize(flow_dag)
+            self._inputs_sign = flow.inputs
+            self._init_sign = flow.init
+        else:
+            # For function entry there is no yaml file to get the inputs and init signature.
+            self._inputs_sign = {}
```
Comment (thy09) [inline src/promptflow-core/promptflow/executor/_script_executor.py:465]: We may add comments to explain what is inputs_sign and init_sign

File: src/promptflow-core/promptflow/executor/flow_validator.py [hunk 2]
```diff
@@ -182,7 +182,9 @@ def resolve_aggregated_flow_inputs_type(flow: Flow, inputs: Mapping[str, List[An
         return updated_inputs
 
     @staticmethod
-    def resolve_flow_inputs_type(flow: Flow, inputs: Mapping[str, Any], idx: Optional[int] = None) -> Mapping[str, Any]:
```
Comment (thy09) [inline src/promptflow-core/promptflow/executor/flow_validator.py:185]: Do not directly change the interface of a public function, we may add another one.

Comment (lumoslnt) [reply src/promptflow-core/promptflow/executor/flow_validator.py]: Refined.

File: src/promptflow-core/promptflow/executor/_script_executor.py (hunk 1 above)
Comment (lumoslnt) [reply src/promptflow-core/promptflow/executor/_script_executor.py:468]: Added.

File: src/promptflow-core/promptflow/executor/_prompty_executor.py [hunk 3]
```diff
@@ -50,3 +50,8 @@ def _initialize_function(self):
         self._inputs = {k: v.to_flow_input_definition() for k, v in inputs.items()}
         self._is_async = False
         return self._func
+
+    def _init_input_sign(self):
```
Comment (thy09) [inline src/promptflow-core/promptflow/executor/_prompty_executor.py:54]: I think prompty also has interfaces, can we clarify whether it is not required?

Comment (lumoslnt) [reply src/promptflow-core/promptflow/executor/_prompty_executor.py:55]: Confirmed with Clement, prompty flow also need to apply default.

Comment (lumoslnt) [reply src/promptflow-core/promptflow/executor/_prompty_executor.py:55]: Confirmed with Clement, SDK wouldn't apply default for prompty. But I found the callable class `Prompty` will do such work in `__call__`. Consider that we should update correct inputs in run info, I prefer applying default value before run info created, update the`_input_sign` here.

**Review by D-W- — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

**Pull request merged**